src/goto-instrument/nondet_volatile.cpp

```
@@ -101,25 +121,102 @@ nondet_volatile(symbol_tablet &symbol_table, goto_programt &goto_program)
+{
+  PRECONDITION(!options.is_set(NONDET_VOLATILE_OPT));
+  PRECONDITION(!options.is_set(NONDET_VOLATILE_VARIABLE_OPT));
+
+  const bool nondet_volatile_opt = cmdline.isset(NONDET_VOLATILE_OPT);
+  const bool nondet_volatile_variable_opt
+    = cmdline.isset(NONDET_VOLATILE_VARIABLE_OPT);
+
+  if(nondet_volatile_opt && nondet_volatile_variable_opt)
+  {
+    throw invalid_command_line_argument_exceptiont(
+      "only one of " NONDET_VOLATILE_OPT "/" NONDET_VOLATILE_VARIABLE_OPT "can "
+      "be given at a time",
+      NONDET_VOLATILE_OPT "/" NONDET_VOLATILE_VARIABLE_OPT);
+  }
+
+  if(nondet_volatile_opt)
+  {
+    options.set_option(NONDET_VOLATILE_OPT, true);
+  }
+  else if(cmdline.isset(NONDET_VOLATILE_VARIABLE_OPT))
+  {
+    options.set_option(
+      NONDET_VOLATILE_VARIABLE_OPT,
+      cmdline.get_values(NONDET_VOLATILE_VARIABLE_OPT));
+  }
+}
+
+void nondet_volatile(goto_modelt &goto_model, const optionst &options)
+{
+  if(options.get_bool_option(NONDET_VOLATILE_OPT))
+  {
+    nondet_volatile(goto_model);
+  }
+  else if(options.is_set(NONDET_VOLATILE_VARIABLE_OPT))
+  {
+    const auto &variable_list =
+      options.get_list_option(NONDET_VOLATILE_VARIABLE_OPT);
+
+    std::set<irep_idt> variables(variable_list.begin(), variable_list.end());
+    const namespacet ns(goto_model.symbol_table);
+
+    // typecheck given variables
+    for(const auto &id : variables)
+    {
+      const symbolt *symbol;
+
+      if(ns.lookup(id, symbol))
+      {
+        throw invalid_command_line_argument_exceptiont(
+          "given name " + id2string(id) + " not found in symbol table",
+          NONDET_VOLATILE_VARIABLE_OPT);
+      }
+
+      if(symbol->is_type || !symbol->type.get_bool(ID_C_volatile))
+      {
+        throw invalid_command_line_argument_exceptiont(
+          "given name " + id2string(id) + " does not represent a volatile "
+          "variable",
+          NONDET_VOLATILE_VARIABLE_OPT);
+      }
+
+      INVARIANT(
```

:question: Indeed - why is this an invariant? Is it possible to trigger this through invalid input - in which case should throw an exception

---

Assuming well-formed goto programs, this should always hold as the type of functions cannot be volatile.